jython-3-mvp.md

```diff
@@ -0,0 +1,236 @@
+
+* Compile Python source to Java byte code, and:
+  * persist compiled Java byte code (in some form).
+  * treat Python compiled to JVM as:
+    * equivalent to cached .pyc files.
+    * resources locatable in a JAR by Jython.
+    * classes visible from Java (maybe).
+
+* Command-line version.
+  * MVP: Launch script or other wrapper. (Generated by `jlink`? C?)
+  * The command you run *is* the interpreter, not a launcher.
+    * JNI appears to [support this readily](
+    https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/invocation.html
+    ).
+    * `sys.executable` designates the actual executable.
+    * Process objects designate the process doing the work (not the wrapper).
+
+* MVP: Interpreter embeddable in a Java application.
+  * Continue JSR-223 support with some clean-up.
+  * `PythonInterpreter` API (only) generally similar to Jython 2.
+  * Risk: current API is unbounded: too much is public.
+    * Reduce the public API aided by Java Jigsaw modules.
+    * There will be blood.
+  * Risk: much existing guidance invalidated (Jython Book).
+
+* Clear semantics for `import` from Java
+  * Semantics in Jython 2 have repeatedly changed.
+  * `import` as written tries the same thing repeatedly
+
+
+## Implementation and API Features
+
+The Java API available for embedding and to extension writers,
+will be heavily influenced by "internal" implementation choices.
+Looked at the other way, premature API choices may constrain implementation
+freedom in undesirable ways.
+
+This is especially true of the object model, since for efficiency's sake,
+objects exchanged at the API will be in the implementations we use internally.
+
+This can be less difficult than in CPython because Java gives us good
+tools for encapsulation: interfaces, packages and modules.
+(Modules are important here. Many implementation details in Jython 2
+became public API, just to cross our own package boundaries.)
+
+* MVP: Resolve the object API before 3.x beta. One of:
+  * `PyObject` is an interface.
+  * Every `j.l.Object` is a Python object directly.
+
+* MVP: Abstract interface (along the lines of CPython `abstract.h`).
+  * Abstract interface for basic operations.
+  * The internals of `type` (slots, etc.) are private
+    (better encapsulated than CPython).
+
+* MVP: Clear relationships amongst interpreter, system state and
+  thread state. (At least the interpreter and its semantics are API.)
+
+
+## Jython 2 Features not to Reproduce
+
+* **Not:** Java `List` and `Map` implicitly Python `list` and `map`.
+  * This is a tempting feature and it almost works,
+    but involves complex guesswork.
+  * Propose a brief way to be explicit instead.
+
+* **Not:** Java package cache.
+  * The location of this is problematic for users.
+  * It has been the subject of a CVE and follow-up bug.
+  * If there is evidence it significantly improves performance,
+    without being a security issue,
+    a reworked version could be added to 3.
+
```

Comment:
Haven't looked into how possible it is but would be nice to avoid use of `jffi`/`jnr` to run completely on top of Java if possible this will make cross platform/architecture support much easier.

Reply:
It's what I meant here:
https://github.com/jython/jython.github.io/pull/21/files#diff-371cb2747f933eaf21de8ddd866028e5R216
Although I may be being too pure.